### DISCUSSION TITLE
[Magick++] smart text alignment for fixed-pt, variable-resolution text

### DISCUSSION THREAD

**Discussion by tsftd** [ImageMagick/ImageMagick#2135] [Help] [answered]
Okay, so I'm wondering if anyone can think of a reasonable way to do this. Essentially, my program:

1) Takes arguments specifying the size/location of the textbox, as well as all font settings
2) Takes a list of text entities to render
3) Automatically determines the largest font size possible such that all text entities will fit within the defined text box
4) creates each text, and outputs a series of images (one for each text entity) with the text centered within the textbox, where each text has the same font size, but different dimensions in pixels.

This all works great, but the one annoying thing is that ascenders and descenders cause the centered text to not align consistently with each other vertically. E.G. if you have 4 words, "can", "dad", "pap", and "dap" -- "can" and "dap" will be (more or less) perfectly centered such that drawing a horizontal line down the center of the textbox will (more or less) bisect both 'a's evenly. However, "dad"'s bisection line will just clip the top of the 'a', while "pap"'s bisection line will just clip the bottom of the 'a'.

The text elements are being correctly centered, but not on the text's baseline, but rather their extremities. This makes it appear as though the text is jumping up and down from one output to the next, if viewed next to each other or in sequence.

Can anyone thing of a (reasonable) way to align them such that the 'a's in all 4 words will be in (more or less) the same vertical position? Without, say, rendering each letter separately and then calculating ascension/descension from the baseline and using that to apply offsets?

**Comment by snibgo** [comment]
You want a number of images, all the same size. Each has some text, all the same pointsize. Is that correct? Then it should work as you want. For example:
```
magick -size 300x200 -pointsize 120 -background #aaf label:dad label:pap +append dadpap.png
```

![image](https://user-images.githubusercontent.com/33812368/84468614-27f9f600-ac77-11ea-9440-ead3ff28a849.png)

**Reply by tsftd** [reply]
Oops, replied to OP instead of your response. Sorry, my response is below.

**Comment by tsftd** [comment]
Um, almost like that. However, it needs to be composited onto a fixed-resolution image, not appended. Also, I'm using Magick++, though in my experience there is *usually* a way to accomplish the same thing as a commandline argument, so if you're more familiar with that, I should be able to work with it. I'm attaching 2 examples: 
![plane](https://user-images.githubusercontent.com/26519234/84469967-37039700-acbd-11ea-9ea0-4b330b774e06.png)
![train](https://user-images.githubusercontent.com/26519234/84469970-3834c400-acbd-11ea-9aaf-f1d9f3e1ea7c.png)

(note: these have been cropped out of the full images and downsampled --this crop is the 'text box' that the text has to be composited on)

The actual code for generating the text (sorry, this forum doesn't seem to handle code blocks/indentation well):
`void compoTXT(struct_layer& llayer,const int_fast32_t subloc,string text,boost::shared_mutex& gate){
	if(text.find("_~")==string::npos){
		for(int_fast32_t lloop=0;lloop<stokens.size();lloop++)while(text.find(stokens[lloop])!=string::npos)text=text.replace(text.find(stokens[lloop]),2,rtokens[lloop]);
		if(text.find("_^")!=string::npos)text=text.substr(0,text.find("_^"))+text.substr(text.find("^_")+2);
		llayer.image[subloc].backgroundColor("transparent");
		llayer.image[subloc].font(llayer.font);
		llayer.image[subloc].fillColor(llayer.color);
		llayer.image[subloc].strokeColor(llayer.color2);
		llayer.image[subloc].fontPointsize(llayer.width);
		llayer.image[subloc].strokeWidth(llayer.outsize*llayer.width);
		llayer.image[subloc].textInterlineSpacing(llayer.interline);
		llayer.image[subloc].textInterwordSpacing(llayer.interword);
		llayer.image[subloc].textKerning(llayer.kerning);
		llayer.image[subloc].textGravity(CenterGravity);
		if(text.find("%")!=string::npos){
			int_fast32_t lloop;
			vector<int_fast32_t> ppos;
			for(lloop=0;lloop<text.size();lloop++)if(text[lloop]=='%')ppos.push_back(lloop);
			while(ppos.size()>0){
				text.insert(ppos.back(),"\\");
				ppos.pop_back();
			}
		}
		llayer.image[subloc].read("label:"+text);
		llayer.image[subloc].colorFuzz(0.1*double(QuantumRange)/100.0);
		bool tflag=true;
		try{llayer.image[subloc].boundingBox();}
		catch(...){tflag=false;}
		if(tflag==true)llayer.image[subloc].trim();
		if(llayer.rotation!=0){
			llayer.image[subloc].rotate(llayer.rotation);
			llayer.image[subloc].colorFuzz(0.1*double(QuantumRange)/100.0);
			tflag=true;
			try{llayer.image[subloc].boundingBox();}
			catch(...){tflag=false;}
			if(tflag==true)llayer.image[subloc].trim();
		}
	}
	else llayer.image[subloc]=dummyimage;
	gate.unlock_shared();
}
`

and for compositing it onto its 'parent layer':

`void compoLYR(struct_layer& llayer,struct_layer& lcanvas,const int_fast32_t subloc,boost::shared_mutex& gate){
	int_fast32_t ltempint,offx=llayer.offx,offy=llayer.offy;
	if(llayer.type=="lyr"&&llayer.rotation!=0){
		llayer.image[subloc].rotate(llayer.rotation);
	}
	if(llayer.type=="lyr"&&llayer.trimobj==true){
		Geometry tempgeo=llayer.image[subloc].boundingBox();
		llayer.image[subloc].colorFuzz(5.0*lexical_cast<double>(QuantumRange)/100.0);
		bool tflag=true;
		try{llayer.image[subloc].boundingBox();}
		catch(...){tflag=false;}
		if(tflag==true)llayer.image[subloc].trim();
		Image tempimage2=llayer.image[subloc];
		llayer.image[subloc]=Image(Geometry(lexical_cast<string>(llayer.image[subloc].baseColumns())+"x"+lexical_cast<string>(llayer.image[subloc].baseRows())),"transparent");
		llayer.image[subloc].composite(tempimage2,tempgeo.xOff(),tempgeo.yOff(),OverCompositeOp);
	}
	if(llayer.gravity==ForgetGravity){
		ltempint=offx;
		while(offy<lcanvas.height-1){
			while(offx<lcanvas.width-1){
				lcanvas.image[0].composite(llayer.image[subloc],offx,offy,OverCompositeOp);
				offx+=llayer.image[subloc].columns();
			}
			offy+=llayer.image[subloc].rows();
			offx=ltempint;
		}
	}
	else{
		if(llayer.gravity==NorthEastGravity||llayer.gravity==EastGravity||llayer.gravity==SouthEastGravity)offx=offx+lcanvas.width-llayer.image[subloc].columns();
		else if(llayer.gravity==NorthGravity||llayer.gravity==SouthGravity||llayer.gravity==CenterGravity)offx=offx+(lcanvas.width-llayer.image[subloc].columns())/2;
		if(llayer.gravity==SouthWestGravity||llayer.gravity==SouthGravity||llayer.gravity==SouthEastGravity)offy=offy+lcanvas.height-llayer.image[subloc].rows();
		else if(llayer.gravity==WestGravity||llayer.gravity==EastGravity||llayer.gravity==CenterGravity)offy=offy+(lcanvas.height-llayer.image[subloc].rows())/2;
		lcanvas.image[0].composite(llayer.image[subloc],offx,offy,OverCompositeOp);
	}
	gate.unlock_shared();
}
`

**Comment by snibgo** [comment, marked as answer]
I appended them to show clearly that the baselines of the two "a" characters align. Is that what you want?

Sorry, I can't read your code. You could put a text file online, and paste a link to it. However, I don't normally debug people's code. But you seem to "trim" the images, which would change the vertical alignment. Your composite(s) also have offsets, which could also change the alignment.

**Reply by tsftd** [reply]
Thank you for the help. The immediate problem was indeed that I was trimming it (I'm not sure if I realized it when I originally wrote the program, but while updating it now I hadn't realized that it would automatically align the text the way I wanted if untrimmed).

However, there was an additional issue: the code for auto-sizing the text doesn't take alignment into consideration, it simply measures the width and height. As such, it is _impossible_ to both maintain identical alignment and fit the text within the target area (e.g. text A is 50px height with +25px above the center and -25px below the center, however text B is 50px height with +15px above the center and -35px below the center -- both texts fit within a target box of 50px height, but if we composite them at the same baseline, either the top of text A or the bottom of text B will be cut off.

What I have done, is add an additional option to either maximize size (the largest pt that will fit in the target area), or preserve alignment (the largest pt that will fit in the target area while keeping standardized alignment). In order to calculate the second, instead of calculating each text separately, it composites all of the texts on top of each other when doing the autosize calculations.

**Comment by tsftd** [comment]
Additional note in case anyone else is trying to do the same thing: to get everything to work properly, I had to add another step. During the automatic size calculation, all the texts are rendered, then composited onto a single canvas, then trimmed. Prior to trimming, save the bounding box for that image.

When you composite out all of the texts onto their respective images, do not trim them -- instead, crop with the saved bounding box. This gives you the ability to "trim" them to a standardized size, and avoids... issues.
